From 58a116e66940c8fca5c34a2e8d6dcce966605941 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Thu, 1 May 2025 12:26:10 +0200
Subject: [PATCH 0923/1013] evl/net: socket: introduce set/get opt requests

With the upcoming support for packet timestamping, we are going to
have more socket options, time has come to provide for dedicated
set/get option requests instead of handling those directly from the
common (inband) ioctl handler.

EVL_SOCIOC_{SET, GET}OPT ioctl codes now redirect to generic
per-socket option handlers. Currently, all options are set/retrieved
from the inband stage, assuming this is merely prep work before oob
I/O transactions take place.

As a result, SENDSZ and RECVSZ ioctl requests have been converted to
socket options.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/uapi/evl/control-abi.h    |   4 +-
 include/uapi/evl/net/socket-abi.h |  27 +++--
 kernel/evl/net/socket.c           | 167 +++++++++++++++++++++++++++---
 3 files changed, 172 insertions(+), 26 deletions(-)

diff --git a/include/uapi/evl/control-abi.h b/include/uapi/evl/control-abi.h
index 03f298b73a3c..386b48360c44 100644
--- a/include/uapi/evl/control-abi.h
+++ b/include/uapi/evl/control-abi.h
@@ -10,14 +10,14 @@
 #include <linux/types.h>
 
 /* Earliest ABI level we support. */
-#define EVL_ABI_BASE   36
+#define EVL_ABI_BASE   39
 /*
  * Current/latest ABI level we support. We may decouple the base and
  * current ABI levels by providing backward compatibility from the
  * latter to the former. CAUTION: a litteral value is required for the
  * current ABI definition (scripts reading this may be naive).
  */
-#define EVL_ABI_LEVEL  38
+#define EVL_ABI_LEVEL  39
 
 #define EVL_CONTROL_DEV  "/dev/evl/control"
 
diff --git a/include/uapi/evl/net/socket-abi.h b/include/uapi/evl/net/socket-abi.h
index 9664c0fb9214..8a8f2060c20d 100644
--- a/include/uapi/evl/net/socket-abi.h
+++ b/include/uapi/evl/net/socket-abi.h
@@ -41,12 +41,25 @@ struct evl_net_solicit {
 	__u32 flags;
 };
 
-#define EVL_SOCKIOC_ACTIVATE	_IOW(EVL_SOCKET_IOCBASE, 2, struct evl_netdev_activation)
-#define EVL_SOCKIOC_DEACTIVATE	_IO(EVL_SOCKET_IOCBASE, 3)
-#define EVL_SOCKIOC_SENDMSG	_IOW(EVL_SOCKET_IOCBASE, 4, struct user_oob_msghdr)
-#define EVL_SOCKIOC_RECVMSG	_IOWR(EVL_SOCKET_IOCBASE, 5, struct user_oob_msghdr)
-#define EVL_SOCKIOC_SETRECVSZ	_IOW(EVL_SOCKET_IOCBASE, 6, int)
-#define EVL_SOCKIOC_SETSENDSZ	_IOW(EVL_SOCKET_IOCBASE, 7, int)
-#define EVL_SOCKIOC_SOLICIT	_IOW(EVL_SOCKET_IOCBASE, 8, struct evl_net_solicit)
+enum {
+	EVL_SOCKOPT_RECVSZ,
+	EVL_SOCKOPT_SENDSZ,
+	EVL_SOCKOPT_TIMESTAMPING,
+};
+
+struct evl_net_sockopt {
+	int level;
+	int option;
+	__u64 optval_ptr;		/* ([const] void __user *optval) */
+	__u64 optlen_ptr;		/* ([const] socklen_t __user *optlen) */
+};
+
+#define EVL_SOCKIOC_ACTIVATE	_IOW(EVL_SOCKET_IOCBASE, 1, struct evl_netdev_activation)
+#define EVL_SOCKIOC_DEACTIVATE	_IO(EVL_SOCKET_IOCBASE, 2)
+#define EVL_SOCKIOC_SENDMSG	_IOW(EVL_SOCKET_IOCBASE, 3, struct user_oob_msghdr)
+#define EVL_SOCKIOC_RECVMSG	_IOWR(EVL_SOCKET_IOCBASE, 4, struct user_oob_msghdr)
+#define EVL_SOCKIOC_SOLICIT	_IOW(EVL_SOCKET_IOCBASE, 5, struct evl_net_solicit)
+#define EVL_SOCKIOC_SETOPT	_IOW(EVL_SOCKET_IOCBASE, 6, struct evl_net_sockopt)
+#define EVL_SOCKIOC_GETOPT	_IOR(EVL_SOCKET_IOCBASE, 7, struct evl_net_sockopt)
 
 #endif /* !_EVL_UAPI_NET_SOCKET_ABI_H */
diff --git a/kernel/evl/net/socket.c b/kernel/evl/net/socket.c
index 4157f9821de3..0e18c993c602 100644
--- a/kernel/evl/net/socket.c
+++ b/kernel/evl/net/socket.c
@@ -572,42 +572,175 @@ __poll_t sock_oob_poll(struct file *filp,
 	return esk->proto->oob_poll(esk, wait);
 }
 
-static int socket_set_rmem(struct evl_socket *esk, int __user *u_val)
+static int get_sockoptaddr(const struct evl_net_sockopt __user *u_opt,
+			void __user **u_optval, unsigned int __user **u_optlen,
+			unsigned int *optlen)
 {
-	int ret, val;
+	__u64 optval_ptr, optlen_ptr;
+	__u32 val;
+	int ret;
+
+	ret = raw_get_user(optval_ptr, &u_opt->optval_ptr);
+	if (ret)
+		return -EFAULT;
+
+	ret = raw_get_user(optlen_ptr, &u_opt->optlen_ptr);
+	if (ret)
+		return -EFAULT;
+
+	*u_optval = evl_valptr64(optval_ptr, void);
+	*u_optlen = evl_valptr64(optlen_ptr, __u32);
+
+	ret = raw_get_user(val, *u_optlen);
+	if (ret)
+		return -EFAULT;
+	if (val < *optlen)
+		return -EINVAL;
+
+	*optlen = val;
+
+	return 0;
+}
+
+static int put_sockopt(void __user *u_optval, const void *optval,
+		unsigned int __user *u_optlen, unsigned int optlen)
+{
+	int ret;
+
+	ret = raw_copy_to_user(u_optval, optval, optlen);
+	if (ret)
+		return ret;
+
+	return raw_put_user(optlen, u_optlen) ? -EFAULT : 0;
+}
+
+static int socket_set_rmem(struct evl_socket *esk,
+			struct evl_net_sockopt __user *u_opt)
+{
+	unsigned int optlen = sizeof(__u32), __user *u_optlen;
+	void __user *u_optval;
+	unsigned int val;
+	int ret;
 
-	ret = raw_get_user(val, u_val);
+	ret = get_sockoptaddr(u_opt, &u_optval,	&u_optlen, &optlen);
 	if (ret)
+		return ret;
+
+	if (raw_get_user(val, (typeof(val) *)u_optval))
 		return -EFAULT;
 
 	/* Same logic as __sock_set_rcvbuf(). */
-	val = min_t(int, val, INT_MAX / 2);
-	WRITE_ONCE(esk->rmem_max, max_t(int, val * 2, SOCK_MIN_RCVBUF));
+	val = min_t(unsigned int, val, UINT_MAX / 2);
+	WRITE_ONCE(esk->rmem_max, max_t(unsigned int, val * 2, SOCK_MIN_RCVBUF));
 
 	return 0;
 }
 
-static int socket_set_wmem(struct evl_socket *esk, int __user *u_val)
+static int socket_get_rmem(struct evl_socket *esk,
+			struct evl_net_sockopt __user *u_opt)
+{
+	unsigned int optlen = sizeof(esk->rmem_max), __user *u_optlen;
+	void __user *u_optval;
+	int ret;
+
+	ret = get_sockoptaddr(u_opt, &u_optval,	&u_optlen, &optlen);
+	if (ret)
+		return ret;
+
+	return put_sockopt(u_optval, &esk->rmem_max,
+			u_optlen, sizeof(esk->rmem_max));
+}
+
+static int socket_set_wmem(struct evl_socket *esk,
+			struct evl_net_sockopt __user *u_opt)
 {
-	int ret, val;
+	unsigned int optlen = sizeof(__u32), __user *u_optlen;
+	void __user *u_optval;
+	unsigned int val;
+	int ret;
 
-	ret = raw_get_user(val, u_val);
+	ret = get_sockoptaddr(u_opt, &u_optval,	&u_optlen, &optlen);
 	if (ret)
+		return ret;
+
+	if (raw_get_user(val, (typeof(val) *)u_optval))
 		return -EFAULT;
 
-	val = min_t(int, val, INT_MAX / 2);
-	WRITE_ONCE(esk->wmem_max, max_t(int, val * 2, SOCK_MIN_SNDBUF));
+	val = min_t(unsigned int, val, UINT_MAX / 2);
+	WRITE_ONCE(esk->wmem_max, max_t(unsigned int, val * 2, SOCK_MIN_SNDBUF));
 
 	return 0;
 }
 
+static int socket_get_wmem(struct evl_socket *esk,
+			struct evl_net_sockopt __user *u_opt)
+{
+	unsigned int optlen = sizeof(esk->wmem_max), __user *u_optlen;
+	void __user *u_optval;
+	int ret;
+
+	ret = get_sockoptaddr(u_opt, &u_optval,	&u_optlen, &optlen);
+	if (ret)
+		return ret;
+
+	return put_sockopt(u_optval, &esk->wmem_max,
+			u_optlen, sizeof(esk->wmem_max));
+}
+
+static int socket_set_option(struct evl_socket *esk,
+			struct evl_net_sockopt __user *u_opt)
+{
+	int level, option;
+
+	if (raw_get_user(level, &u_opt->level))
+		return -EFAULT;
+
+	if (level != SOL_SOCKET)
+		return -EINVAL;
+
+	if (raw_get_user(option, &u_opt->option))
+		return -EFAULT;
+
+	switch (option) {
+	case EVL_SOCKOPT_RECVSZ:
+		return socket_set_rmem(esk, u_opt);
+	case EVL_SOCKOPT_SENDSZ:
+		return socket_set_wmem(esk, u_opt);
+	}
+
+	return -EINVAL;
+}
+
+static int socket_get_option(struct evl_socket *esk,
+			struct evl_net_sockopt __user *u_opt)
+{
+	int level, option;
+
+	if (raw_get_user(level, &u_opt->level))
+		return -EFAULT;
+
+	if (level != SOL_SOCKET)
+		return -EINVAL;
+
+	if (raw_get_user(option, &u_opt->option))
+		return -EFAULT;
+
+	switch (option) {
+	case EVL_SOCKOPT_RECVSZ:
+		return socket_get_rmem(esk, u_opt);
+	case EVL_SOCKOPT_SENDSZ:
+		return socket_get_wmem(esk, u_opt);
+	}
+
+	return -EINVAL;
+}
+
 static long sock_inband_ioctl(struct sock *sk, unsigned int cmd,
 			unsigned long arg)
 {
 	struct evl_socket *esk = evl_sk(sk);
 	struct evl_netdev_activation act, __user *u_act;
 	struct evl_net_solicit solreq, __user *u_solreq;
-	int __user *u_val;
 	int ret;
 
 	switch (cmd) {
@@ -621,13 +754,13 @@ static long sock_inband_ioctl(struct sock *sk, unsigned int cmd,
 	case EVL_SOCKIOC_DEACTIVATE: /* Turn oob port off. */
 		ret = evl_net_switch_oob_port(esk, NULL);
  		break;
-	case EVL_SOCKIOC_SETRECVSZ:
-		u_val = (typeof(u_val))arg;
-		ret = socket_set_rmem(esk, u_val);
+	case EVL_SOCKIOC_SETOPT:
+		ret = socket_set_option(esk,
+			(struct evl_net_sockopt __user *)arg);
 		break;
-	case EVL_SOCKIOC_SETSENDSZ:
-		u_val = (typeof(u_val))arg;
-		ret = socket_set_wmem(esk, u_val);
+	case EVL_SOCKIOC_GETOPT:
+		ret = socket_get_option(esk,
+			(struct evl_net_sockopt __user *)arg);
 		break;
 	case EVL_SOCKIOC_SOLICIT:
 		u_solreq = (typeof(u_solreq))arg;
-- 
2.30.2

